### PR TITLE
[T4MVC] attribute

### PR DESCRIPTION
**Pull request by migig**
added [T4MVC] attribute for controllers
removed obsolete "ControllerNamePattern" from T4MVC.tt.settings.xml and T4MVC.tt
added some controller test cases

as discussed:
- backwards compatible... you don't need to use this attribute if you don't want to
- supports regular controllers (i.e. backwards compatible)
- supports abstract controllers
- supports weirdly named controllers

The generic controllers stuff is excluded. We can work on that separately.

As mentioned in another issue, it won't build out the box as it's missing a ref to System.Web.Mvc.dll. I didn't add that to the commit, as I only wanted to focus on my stuff. The template works though, it's only the sample project that doesn't build...

If you need documentation, how about:

**[T4MVC] attribute:**
By default, T4MVC ignores abstract controllers, and controllers which do not follow the naming convention of `FooController`. If you'd like to process such controllers, decorate the class with the `[T4MVC]` attribute. You can also force T4MVC to ignore a controller (which T4MVC automatically processes) by decorating it with the `[T4MVC(false)]` attribute.

### CODE REVIEW

**Comment by migig** [comment]
BTW there's a class full of controller test cases.

File: T4MVCHostMvcApp/Controllers/TestAttributeControllers.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  // look for the results under the T4MVC.tt node
+  // those with "YES" should appear, those with "NO" should not
+  //
+
+
+
+  /// test regular controllers' names ...................................................................
+
+  // this controller should be processed, and its "Controller" substring removed
+  public partial class TestAttribute_YES_Name1Controller : Controller { }
```
Comment (davidebbo) [inline T4MVCHostMvcApp/Controllers/TestAttributeControllers.cs]: Getting a lot of Code Analysis violations in this file. e.g. this line causes:

    CA1707	Identifiers should not contain underscores	Remove the underscores from type name 'TestAttribute_YES_Name1Controller'.

Search for `SuppressMessage` in the sources to see how this is handled.

Comment (davidebbo) [inline T4MVCHostMvcApp/Controllers/TestAttributeControllers.cs]: BTW, I get a total of 21 errors like this. Not sure why you don't see them. Maybe Code Analysis only happens in some versions of VS.

File: T4MVCHostMvcApp/T4MVC Files/T4MVC.tt [hunk 2]
```diff
@@ -751,7 +752,11 @@ void ProcessControllerType(CodeClass2 type, AreaInfo area, DateTime controllerLa
     if (type.IsAbstract)
     {
         // If it's abstract, set a flag and don't process action methods (derived classes will)
-        target.IsAbstract = true;
+        if (!HasControllerAttribute(type))
+            target.IsAbstract = true;
+        // ...unless it has the [T4MVC] attribute, then process all the action methods in the controller
+        else
+          ProcessControllerActionMethods(target, type);
```
Comment (davidebbo) [inline T4MVCHostMvcApp/T4MVC Files/T4MVC.tt]: The style we use in T4MVC is 4 spaces, while you have 2 in a bunch of places (in this file and others).

**Comment by davidebbo** [comment]
Change looks great overall. Just minor comments about spacing and suppressing Code Analysis issues and it's good to go. Thanks!

**Comment by migig** [comment]
First time use of Code Analysis tool. Changes made as requested.

**Comment by davidebbo** [comment]
Ok, it's in and pushed as 3.12.0!

**Comment by migig** [comment]
Cool!

I wrote up some documentation above in the original text.. maybe include it in the official docs? Otherwise I'm the only one to benefit from the feature :)

**Comment by migig** [comment]
BTW I'll open a separate issue to deal with the generic controllers case, as well as my first stab at some code that ALMOST works.

**Comment by migig** [comment]
Nope there's an [unexpected problem](https://github.com/T4MVC/T4MVC/issues/6).

**Pull request closed without merging**